storage/index.go

```diff
@@ -46,50 +60,120 @@ const (
+// lazily updated, so it can point to an older element until a Tick()/write rotates it.
+// nolint: maligned
 type nsIndex struct {
-	sync.RWMutex
-	insertMode index.InsertMode
-	state      nsIndexState
-	active     index.Block
+	sync.RWMutex         // NB: this mutex guards all fields until (and including) latestBlock
+	state                nsIndexState
+	bootstrapState       BootstrapState
+	insertQueue          namespaceIndexInsertQueue
+	runtimeOpts          nsIndexRuntimeOptions
+	blocksByTime         map[xtime.UnixNano]index.Block
+	blockStartsDescOrder []xtime.UnixNano
+	latestBlock          index.Block
+
+	// all the vars below this line are not modified past the ctor
+	// and don't require a lock when being accessed.
+	nowFn           clock.NowFn
+	blockSize       time.Duration
+	retentionPeriod time.Duration
+	bufferPast      time.Duration
+	bufferFuture    time.Duration
+
+	newBlockFn newBlockFn
+	logger     xlog.Logger
+	opts       index.Options
+	metrics    nsIndexMetrics
+	nsMetadata namespace.Metadata
+}
 
-	insertQueue namespaceIndexInsertQueue
-	metrics     nsIndexMetrics
-	opts        index.Options
-	nsID        ident.ID
+// NB: nsIndexRuntimeOptions does not contain its own mutex as some of the variables
+// are needed for each index write which already at least acquires read lock from
+// nsIndex mutex, so to keep the lock acquisitions to a minimum these are protected
+// under the same nsIndex mutex.
+type nsIndexRuntimeOptions struct {
+	insertMode    index.InsertMode
+	maxQueryLimit int64
 }
 
+type newBlockFn func(time.Time, time.Duration, index.Options) (index.Block, error)
+
 func newNamespaceIndex(
-	md namespace.Metadata,
+	nsMD namespace.Metadata,
+	opts index.Options,
+) (namespaceIndex, error) {
+	return newNamespaceIndexTestCtor(nsMD, newNamespaceIndexInsertQueue, index.NewBlock, opts)
+}
+
+// newNamespaceIndexWithInsertQueueFn is a ctor used in tests to override the insert queue.
+// nolint: deadcode
+func newNamespaceIndexWithInsertQueueFn(
+	nsMD namespace.Metadata,
 	newIndexQueueFn newNamespaceIndexInsertQueueFn,
 	opts index.Options,
 ) (namespaceIndex, error) {
-	if err := opts.Validate(); err != nil {
-		return nil, err
-	}
+	return newNamespaceIndexTestCtor(nsMD, newIndexQueueFn, index.NewBlock, opts)
+}
 
-	now := opts.ClockOptions().NowFn()()
-	indexBlockSize := md.Options().IndexOptions().BlockSize()
-	start := now.Truncate(indexBlockSize)
-	blk, err := index.NewBlock(start, indexBlockSize, opts)
-	if err != nil {
+// newNamespaceIndexWithNewBlockFn is a ctor used in tests to inject blocks.
+// nolint: deadcode
+func newNamespaceIndexWithNewBlockFn(
+	nsMD namespace.Metadata,
+	newBlockFn newBlockFn,
+	opts index.Options,
+) (namespaceIndex, error) {
+	return newNamespaceIndexTestCtor(nsMD, newNamespaceIndexInsertQueue, newBlockFn, opts)
+}
+
+// newNamespaceIndexTestCtor is a ctor used in tests to inject ctor functions for block/queue.
+func newNamespaceIndexTestCtor(
```

Comment:
This is also used by the prod code path yeah as `newNamespaceIndex(...)` calls this.  Perhaps a better name is `newNamespaceIndexWithOptions(...)` and make a `namespaceIndexOptions` structs with the dependencies on it?

Reply:
Fair. The comment was directed at external (outside the file) callers but it's misleading, will re-work as suggested.

Reply:
done